commit/merkleroot/rmn/client.go

```diff
@@ -19,13 +33,533 @@ type Client interface {
+		lggr:                                    lggr,
+		rawRmnClient:                            rawRmnClient,
+		rmnNodes:                                rmnNodes,
+		rmnRemoteAddress:                        rmnRemoteAddress,
+		rmnHomeConfigDigest:                     rmnHomeConfigDigest,
+		minObservers:                            minObservers,
+		minSigners:                              minSigners,
+		observationsInitialRequestTimerDuration: observationsInitialRequestTimerDuration,
+		reportsInitialRequestTimerDuration:      reportsInitialRequestTimerDuration,
+	}
+}
+
+func (c *PBClient) ComputeReportSignatures(
+	ctx context.Context,
+	destChain *rmnpb.LaneDest,
+	requestedUpdates []*rmnpb.FixedDestLaneUpdateRequest,
+) (*ReportSignatures, error) {
+	// Group the lane update requests by their source chain.
+	lursPerChain := make(map[uint64]lurWithMeta)
+	for _, updateReq := range requestedUpdates {
+		if _, exists := lursPerChain[updateReq.LaneSource.SourceChainSelector]; exists {
+			return nil, errors.New("this Client implementation assumes each lane update is for a different chain")
+		}
+
+		lursPerChain[updateReq.LaneSource.SourceChainSelector] = lurWithMeta{
+			lur:      updateReq,
+			rmnNodes: mapset.NewSet[uint32](),
+		}
+		for _, node := range c.rmnNodes {
+			if node.SupportedSourceChains.Contains(updateReq.LaneSource.SourceChainSelector) {
+				lursPerChain[updateReq.LaneSource.SourceChainSelector].rmnNodes.Add(node.ID)
+			}
+		}
+	}
+
+	// Filter out the lane update requests for chains without enough RMN nodes supporting them.
+	for chain, l := range lursPerChain {
+		if l.rmnNodes.Cardinality() < c.minObservers {
+			delete(lursPerChain, chain)
+		}
+	}
+
+	if len(lursPerChain) == 0 {
+		return nil, errors.New("no source chains with enough RMN nodes, nothing to do")
+	}
+
+	rmnSignedObservations, err := c.getRmnSignedObservations(ctx, destChain, lursPerChain)
+	if err != nil {
+		return nil, fmt.Errorf("get rmn signed observations: %w", err)
+	}
+
+	rmnReportSignatures, err := c.getRmnReportSignatures(ctx, destChain, rmnSignedObservations)
+	if err != nil {
+		return nil, fmt.Errorf("get rmn report signatures: %w", err)
+	}
+
+	return rmnReportSignatures, nil
+}
+
+func (c *PBClient) getRmnSignedObservations(
+	ctx context.Context,
+	destChain *rmnpb.LaneDest,
+	lursPerChain map[uint64]lurWithMeta,
+) ([]rmnSignedObservationWithMeta, error) {
+	requestedNodes := make(map[uint64]mapset.Set[uint32])                   // sourceChain -> requested rmnNodeIDs
+	requestsPerNode := make(map[uint32][]*rmnpb.FixedDestLaneUpdateRequest) // grouped requests for each node
+
+	// For each lane update request send an observation request to at most 'minObservers' number of rmn nodes.
+	for sourceChain, lur := range lursPerChain {
+		requestedNodes[sourceChain] = mapset.NewSet[uint32]()
+
+		// At this point we assume at least minObservers RMN nodes are available for each source chain.
+		for nodeID := range lur.rmnNodes.Iter() {
```

Comment:
I think we should shuffle this node list prior to iterating and choosing the RMN node to send to right?

Reply:
I think you can use `randomIter`, seems like you use it somewhere else just not here.

Reply:
I thought Iter() would iterate over the elements randomly because it's a set, but it doesn't, I am curious if mapset really persists ordering.

Changed it with randomIter(rmnNodes.ToSlice())